packages/ai/src/generate-text/generate-text.ts

```diff
@@ -678,11 +693,41 @@ A function that attempts to repair a tool call that failed to parse.
             toolApprovalRequests: Object.values(toolApprovalRequests),
           });
 
+          // If a provider-executed approval was denied, add a UI-visible closure without
+          // persisting it into the returned response messages (so it is not sent to providers).
+          const excludedToolResultToolCallIds = new Set<string>();
+          if (pendingDeniedProviderExecutedToolApprovals.length > 0) {
+            for (const deniedApproval of pendingDeniedProviderExecutedToolApprovals) {
+              excludedToolResultToolCallIds.add(
+                deniedApproval.toolCall.toolCallId,
+              );
+              stepContent.push({
+                type: 'tool-result',
+                toolCallId: deniedApproval.toolCall.toolCallId,
+                toolName: deniedApproval.toolCall.toolName,
+                input: deniedApproval.toolCall.input,
+                output: {
+                  type: 'execution-denied',
+                  reason: deniedApproval.approvalResponse.reason,
+                },
+                providerExecuted: true,
+                dynamic: true,
+              });
+            }
+
+            // ensure we only emit this closure once:
+            pendingDeniedProviderExecutedToolApprovals = [];
+          }
```

Comment:
this code is strange - what do we emit here? why is the if needed (and not just the loop)?